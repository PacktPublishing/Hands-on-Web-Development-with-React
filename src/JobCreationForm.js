import React from 'react';

export default class JobCreationForm extends React.Component {
  render() {
    return (
      <form>
      </form>
    );
  }
}
