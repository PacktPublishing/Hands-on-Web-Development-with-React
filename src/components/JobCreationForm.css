.job-form {
  width: 80%;
  margin: 0 auto;
  box-shadow: 0px 2px 40px 0 rgba(0, 0, 0, 0.25);
  padding: 24px 12px;
}

.job-form__button {
  font-size: 1.25em;
  padding: 4px 8px;
}

.job-form__error-label {
  color: crimson;
}
