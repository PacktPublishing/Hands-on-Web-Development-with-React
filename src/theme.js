export default {
  colors: {
    primary: 'royalblue',
    secondary: 'crimson',
    text: 'gray',
  }
};
