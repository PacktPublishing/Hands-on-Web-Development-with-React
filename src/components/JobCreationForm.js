import React from 'react';
import TextInputField from './form-elements/TextInputField';
import CheckboxField from './form-elements/CheckboxField';
import SelectField from './form-elements/SelectField';
import ConsentInput from './form-elements/ConsentInput';
import { PrimaryButton } from './Button';
import './JobCreationForm.css';
import Spinner from './Spinner';
import isFormDataValid from '../utils/jobCreationFormValidator';

const locationOptions = [
  {value: '', label: 'Blank' },
  {value: 'Berlin', label: 'Berlin' },
  {value: 'San Francisco', label: 'San Francisco' },
  {value: 'London', label: 'London' },
  {value: 'Austin', label: 'Austin' },
  {value: 'Tokyo', label: 'Tokyo' },
  {value: 'Barcelona', label: 'Barcelona' },
  {value: 'Other', label: 'Other' },
];

const initialState = {
  title: '',
  company: '',
  salary: '',
  isRemoteFriendly: false,
  location: '',
  acceptedToS: false,
  subscribedToNewsletter: false,
};

export default class JobCreationForm extends React.Component {
  state = initialState;

  componentDidMount = () => {
    if (this.props.defaultState) {
      const defaultState = {};
      Object.keys(initialState).forEach((key) => {
        defaultState[key] = this.props.defaultState[key] || initialState[key];
      });
      this.setState(defaultState);
    }
  };

  handleChange = (e) => {
    const { type, name, value, checked } = e.target;
    if (type === 'checkbox') {
      this.setState({ [name]: checked });
    } else {
      this.setState({ [name]: value });
    }
    if (name === 'title' && value.length >= 10) {
      this.setState({ titleError: false });
    }
  };

  handleSubmit = (e) => {
    e.preventDefault();
    this.props.onSubmit(this.state);
  };

  handleBlur = (e) => {
    const { name } = e.target;
    if (name === 'title') {
      this.setState({ titleError: this.state.title.length < 10 });
    }
  };

  render() {
    const enabled = isFormDataValid(this.state);
    return (
      <form
        className="job-form"
        onSubmit={this.handleSubmit}
      >
        <TextInputField
          name="title"
          label="Title"
          onChange={this.handleChange}
          value={this.state.title}
          required={true}
          onBlur={this.handleBlur}
        />
        {this.state.titleError ?
          <p className="job-form__error-label">
            Should be at least 10 characters long.
          </p> :
          null
        }
        <TextInputField
          name="company"
          label="Company"
          onChange={this.handleChange}
          value={this.state.company}
          required={true}
        />
        <TextInputField
          name="salary"
          label="Salary"
          onChange={this.handleChange}
          value={this.state.salary}
          required={true}
        />
        <CheckboxField
          name="isRemoteFriendly"
          label="Remote friendly?"
          onChange={this.handleChange}
          value={this.state.isRemoteFriendly}
        />
        <SelectField
          name="location"
          label="Location"
          options={locationOptions}
          onChange={this.handleChange}
          value={this.state.location}
        />
        <ConsentInput
          onChange={this.handleChange}
          acceptedToS={this.state.acceptedToS}
          subscribedToNewsletter={this.state.subscribedToNewsletter}
        />
        <PrimaryButton disabled={!enabled || this.props.isSubmitting}>
          Submit Job
        </PrimaryButton>
        {this.props.isSubmitting &&
          <Spinner />
        }
      </form>
    );
  }
}
