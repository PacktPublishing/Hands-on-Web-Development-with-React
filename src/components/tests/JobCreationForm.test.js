import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import JobCreationForm from '../JobCreationForm';
import theme from '../../theme';

const log = (e) => console.log(e);

describe('JobCreationForm.js', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={theme}>
          <JobCreationForm onSubmit={log} />
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
