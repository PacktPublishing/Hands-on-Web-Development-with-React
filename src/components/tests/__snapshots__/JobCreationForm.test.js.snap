// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobCreationForm.js renders correctly 1`] = `
<form
  className="job-form"
  onSubmit={[Function]}
>
  <div
    className="input-field"
  >
    <label
      className="input-field__label"
      htmlFor="title"
    >
      Title
    </label>
    <input
      className="input-field__text-input"
      id="title"
      name="title"
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="input-field"
  >
    <label
      className="input-field__label"
      htmlFor="company"
    >
      Company
    </label>
    <input
      className="input-field__text-input"
      id="company"
      name="company"
      onBlur={undefined}
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="input-field"
  >
    <label
      className="input-field__label"
      htmlFor="salary"
    >
      Salary
    </label>
    <input
      className="input-field__text-input"
      id="salary"
      name="salary"
      onBlur={undefined}
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="checkbox-field"
  >
    <input
      id="isRemoteFriendly"
      name="isRemoteFriendly"
      onChange={[Function]}
      type="checkbox"
      value={false}
    />
    <label
      htmlFor="isRemoteFriendly"
    >
      Remote friendly?
    </label>
  </div>
  <div
    className="input-field"
  >
    <label
      className="input-field__label"
      htmlFor="location"
    >
      Location
    </label>
    <select
      className="input-field__text-input"
      id="location"
      name="location"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        Blank
      </option>
      <option
        value="Berlin"
      >
        Berlin
      </option>
      <option
        value="San Francisco"
      >
        San Francisco
      </option>
      <option
        value="London"
      >
        London
      </option>
      <option
        value="Austin"
      >
        Austin
      </option>
      <option
        value="Tokyo"
      >
        Tokyo
      </option>
      <option
        value="Barcelona"
      >
        Barcelona
      </option>
      <option
        value="Other"
      >
        Other
      </option>
    </select>
  </div>
  <div>
    <div
      className="checkbox-field"
    >
      <input
        id="acceptedToS"
        name="acceptedToS"
        onChange={[Function]}
        type="checkbox"
        value={false}
      />
      <label
        htmlFor="acceptedToS"
      >
        Accept ToS
      </label>
    </div>
    <div
      className="checkbox-field"
    >
      <input
        id="subscribedToNewsletter"
        name="subscribedToNewsletter"
        onChange={[Function]}
        type="checkbox"
        value={false}
      />
      <label
        htmlFor="subscribedToNewsletter"
      >
        Send me your newsletter!
      </label>
    </div>
  </div>
  <button
    className="sc-bwzfXH biIyjb"
    disabled={true}
  >
    Submit Job
  </button>
</form>
`;
